### ISSUE TITLE
how to not display entire row if there is one null value in a column?

### ISSUE THREAD

**Issue by oscar6654** [jbox-web/ajax-datatables-rails#163]
Im trying to do .strftime but if the data in that specific record does not have a value, strftime will give out an error. so how do i just remove the entire row if there is no data in that record that strftime is trying to convert?

**Comment by oscar6654** [comment]
i managed to make it work by doing. 

class NilClass
  def strftime(_)
    ''
  end
end

I need to know how to not show records that are empty 
so for example
xxxxxxxxxxxxxxxxxxxxxxxxx
Invoice Num | Invoice Date
1231312     | 2016-7-14
3123123     |
xxxxxxxxxxxxxxxxxxxxxxxxx
then the table should hide 3123123's row because invoice date is null or empty so it wont render it.

**Comment by ajahongir** [comment]
@oscar6654 in datatable you have to define:

``` ruby
def data
    records.map do |invoice|
      {
        ...
        date: (invoice.date.present? ? invoice.date.strftime('%Y-%m-%d') : nil),
        ....
      }
    end
  end
```

**Comment by oscar6654** [comment]
yes that worked but is there a possible way not to display them in the table ?  I can do the if statement but dont know how it works.
so it be:
if invoice.date != nil
{
invoice.name
invoice.item
....
}

**Comment by oscar6654** [comment]
like an example i have is this.. but its not using ajax just normal way. 
 <tbody>
      <% @invoices.each do |user| %>
      <%- if user.post_status != "Closed" && user.delivery_status != "Deliver"%>
      <tr>
        <td><%= check_box_tag "invoice_ids[]", user.id %></td>
        <td><%= user.invoicenum %></td>
        <td><%= Seller.find(user.sellername.to_i).name%></td>
        <td><%= user.account_name %></td>
        <td><%= user.whlocation %></td>
        <td><%= number_to_currency(user.invoice_amnt, :locale => :en) %></td>
        <td><%= (user.invoice_created_at).strftime("%m/%d/%Y") %></td>
        <td><%= user.delivery_status %></td>
        <%- if user.delivery_date == nil %>
        <td><%= "No Requested Delivery Date" %></td>
        <%-else%>
        <td><%= user.delivery_date %></td>
        <%end%>
        <td><%= user.post_status %></td>
      </tr>
      <%end%>
      <%end%>
    </tbody>
</table>

so basically all post_status that is closed does not get displayed at all... so for date all date that are  not nil does not get displayed including other items..

**Comment by ajahongir** [comment]
if you do not want to show a row if data is an empty:

``` ruby
def records
  Invoice.where('date <> NULL')
end
```

**Comment by oscar6654** [comment]
i did that then now shows.. no matching records found

**Comment by ajahongir** [comment]
can you show your datatable?

**Comment by oscar6654** [comment]
here is the screenshot example: http://imgur.com/OJZxFgo
so I do not want to show  Oscar Chng23 entire row since dispatch datetime field is null.
this is my datatable:

class PlateDatatable < AjaxDatatablesRails::Base
  include AjaxDatatablesRails::Extensions::Kaminari
  def_delegator :@view, :link_to
  def_delegator :@view, :number_to_currency
  def sortable_columns
    # Declare strings in this format: ModelName.column_name
    @sortable_columns ||= ['invoices.plate_no','invoices.dispatch_datetime','invoices.account_name','invoices.invoicenum','invoices.invoice_amnt','invoices.delivery_crew']
    #@sortable_columns ||= ['invoices.invoicenum','invoices.sellername','invoices.account_name','invoices.whlocation','invoices.invoice_amnt','invoices.invoice_created_at','invoices.delivery_status','invoices.delivery_date','invoices.post_status']
  end

  def searchable_columns
    # Declare strings in this format: ModelName.column_name
    @searchable_columns ||= ['invoices.plate_no','invoices.dispatch_datetime','invoices.account_name','invoices.invoicenum','invoices.invoice_amnt','invoices.delivery_crew']
    #@searchable_columns ||= ['invoices.invoicenum','invoices.sellername','invoices.account_name','invoices.whlocation','invoices.invoice_amnt','invoices.invoice_created_at','invoices.delivery_status','invoices.delivery_date','invoices.post_status']
  end

  private

  def data
    records.map do |record|

```
   [
    record.plate_no,
    #record.dispatch_datetime.strftime('%Y-%m-%d'),
    (record.dispatch_datetime.present? ? record.dispatch_datetime.strftime('%Y-%m-%d') : nil),
    record.account_name,
    record.invoicenum,
    number_to_currency(record.invoice_amnt),
    record.delivery_crew


  #   # comma separated list of the values for each cell of a table row
  #   # example: record.attribute,
   ]
```

  end
  end

  def get_raw_records
    # insert query here
    Invoice.all

  end

  def records
    #Invoice.where('plate_no <> NULL')
  end
  # ==== Insert 'presenter'-like methods below if necessary
end

**Comment by ajahongir** [comment]
def get_raw_records
  Invoice.where('plate_no <> NULL')
end

**Comment by oscar6654** [comment]
i did that, and now it says no data available in table

**Comment by ajahongir** [comment]
Invoice.where('plate_no IS NOT NULL') ?
are you sure that in db your columns are null and not an empty string?

**Comment by oscar6654** [comment]
they are null, not empty strings

**Comment by oscar6654** [comment]
ok the IS NOT NULL worked. thanks!

**Comment by antillas21** [comment]
Yes, if you want to skip a complete record from the display, the corresponding code should be in the query to your database.

I know this is already solved, but if you wanted to go in a more ruby-like approach:

``` ruby
# rails 4+
def get_raw_records
  Invoice.where.not(plate_no: nil)
end

# rails 3
def get_raw_records
  Invoice.where("plate_no is not null")
end
# I don't recall if you can resort to Arel to solve this
# in this version of rails.
# I mean, Arel is there, but don't remember if
# in this version, it supports `not`
```
